### DISCUSSION TITLE
Global extension methods

### DISCUSSION THREAD

**Discussion by nikolazlatkovic** [dotnet/csharplang#6183] [Language Ideas] [answered]
Are there any reasons to have extension methods reside inside of (static) classes, when they already do have defined a "receiver" type, which they operate on?

To illustrate, here is a (working) "dummy" example:
```csharp
var list = Enumerable.Range(1, 2).ToList2();

public static class EnumerableExtensions
{
    public static IList<T> ToList2<T>(this IEnumerable<T> source) => 
        source.ToList();
}
```
In this case, class `EnumerableExtensions` is not being referenced anywhere.

There are lots of similar examples, even in .NET codebase, where classes named `<<TypeName>>Extensions` have been created, only to serve as "containers" for extension methods.

Good examples are ASP.NET Core Middleware implementations, where each one has own `ServiceCollectionExtensions` class just to contain `UseSomeMiddleware(this IApplicationBuilder app)` method(s).

Allowing them to reside directly inside of namespace, in a similar manner to [extension functions](https://kotlinlang.org/docs/extensions.html#extension-functions) in Kotlin, should not break any existing code, yet it could make it even more concise, alongside all other efforts to make it so.

Just a food for thought...

**Comment by HaloFour** [comment, marked as answer]
Extension methods are just syntax candy over a normal static method.  They need to be directly invocable as a static method.

Extension functions in Kotlin also exist in a class, but the Kotlin language gives that class a name automatically based on the name of the source file.  But that's partly because in the JVM world the source file name has a direct relationship with the class being compiled, whereas that relationship does not exist in C#.  The team has also generally shied away from having the compiler automatically generate names for things that are exposed publicly as the method through which the name is synthesized would have to be described in the C# specification so that changing between compilers doesn't cause a binary breaking change.

**Reply by HaloFour** [reply]
Also check out #5496 which reimagines the syntax for extensions.  Still very early in design, and would still necessitate naming the extension class, but would allow for more kinds of extension members like properties and potentially interface implementation.
